src/librustc_mir/borrow_check/nll/region_infer.rs

```diff
@@ -423,7 +479,49 @@ impl<'tcx> RegionInferenceContext<'tcx> {
             }
 
             let point_index = self.point_indices.get(&p).unwrap();
-            changed |= inferred_values.add(to_region.index(), *point_index);
+            let new = inferred_values.add(to_region.index(), *point_index);
+            changed |= new;
+            if new {
+                let re = if self.definitions[from_region].name.is_some() {
+                    RegionElement::FreeRegion(from_region)
+                } else {
+                    RegionElement::Location(p)
+                };
+                let statements = &mir.basic_blocks()[p.block].statements;
+                let span = if p.statement_index < statements.len() {
+                    statements[p.statement_index].source_info.span
+                } else {
+                    DUMMY_SP
+                };
+                // TODO(pfaria) Find more efficient version that doesn't require cloning (split_at_mut?)
+                let new_causes = causes[from_region].iter()
```

Comment:
so at this point we know that:

- `from_region` contains the point `p`
- we are adding `p` into `to_region`
- so we can get the old cause `old_clause` via `self.causes[(from_region, p)]`
- and construct a new cause `Cause::Subregion { ... original_clause: old_clause }`